data/index.ts

```diff
@@ -45,69 +46,99 @@ function setTop(benefit, institution) {
-  benefit.top = setTop(benefit, institution)
-  benefit.floorAt = benefit.floorAt || 1
-  return benefit
-}
+function generateFullBenefitData(benefitCollection, institutions) {
+  benefitCollection.map((benefit) => {
+    benefit.id = generateBenefitId(benefit)
+    benefit.top = setTop(benefit, institutions[benefit.institution])
+    benefit.floorAt = benefit.floorAt || 1
+    return benefit
+  })
 
-export function generate(
-  collections,
-  additionalBenefitAttributes,
-  aidesVeloBenefitListGenerator
+  return benefitCollection
+}
+const mergeBenefits = function (
+  staticBenefitCollection,
+  institutions,
+  options
 ) {
-  const institutions = transformInstitutions(collections.institutions.items)
+  const benefitCollection = [
+    ...staticBenefitCollection.benefits_javascript.items,
+    ...staticBenefitCollection.benefits_openfisca.items,
+  ]
 
-  collections.benefits_javascript.items.forEach((benefit) => {
+  staticBenefitCollection.benefits_javascript.items.forEach((benefit) => {
     benefit.source = "javascript"
   })
-  collections.benefits_openfisca.items.forEach((benefit) => {
+  staticBenefitCollection.benefits_openfisca.items.forEach((benefit) => {
     benefit.source = "openfisca"
   })
 
-  const aidesVeloBenefits = aidesVeloBenefitListGenerator
-    ? aidesVeloBenefitListGenerator(Object.values(institutions))
-    : []
-  aidesVeloBenefits.forEach((benefit) => {
-    benefit.source = "aides-velo"
-  })
-
-  const fslBenefits = build(institutions)
-
-  let benefits = [
-    ...collections.benefits_javascript.items,
-    ...collections.benefits_openfisca.items,
-    ...aidesVeloBenefits.filter((b) => b.institution),
-    ...fslBenefits,
-  ].map((benefit) => {
-    return Object.assign({}, benefit, additionalBenefitAttributes[benefit.slug])
+  if (options.fsl === true) {
+    benefitCollection.push(...buildFSL(institutions))
+  }
+  if (options.aidesVelo === true) {
+    const aidesVeloBenefits = aidesVeloGenerator(Object.values(institutions))
+    aidesVeloBenefits.map((benefit: benefitVeloLayout) => {
+      benefit.source = "aides-velo"
+    })
+    benefitCollection.push(...aidesVeloBenefits)
+  }
+  benefitCollection.map((benefit) => {
+    return Object.assign(
+      benefit,
+      options.dynamicBenefitAttributes[benefit.slug]
+    )
   })
 
-  const benefitsMap = {}
-
-  benefits = benefits.map((benefit) => {
+  return benefitCollection
+}
+const addRelationToData = function (institutions, benefitCollection) {
+  benefitCollection.forEach((benefit) => {
     const institution = institutions[benefit.institution]
-    benefit = setDefaults(benefit, institution)
     institution.benefitsIds.push(benefit.id)
     benefit.institution = institution
+  })
+}
+export function generate(
+  staticBenefitCollection,
+  options = {
+    // permet de déterminer les infos à générer pour les tests unitaires
+    fsl: true,
+    aidesVelo: true,
+    dynamicBenefitAttributes: dynamicBenefitAttributes,
+  }
+) {
```

Comment:
Est ce que FSL et AidesVelo pourraient être gérer via une `dynamicBenefitGenerators` liste qui serait un paramètre, et sur laquelle on itérerait. Plutôt que des flags avec des booléans ?

Reply:
je suis pas sûr de voir la plus-value. 

A la rigueur sortir dynamicBenefit c'est pas aberrant (d'autant que ça sert que pour 1 test, et idéalement faudrait supprimer l'argument et avoir des données fictives plutôt que de les injecter a l'appel de la méthode) 

mais le dictionnaire de boolean permet d'avoir tous les builders dans index et d'avoir une liste exhaustive de ce qu'on peut appeler au moment du call

Reply:
Du coup j'ai coupé la poire en deux, 
```javascript
  staticBenefitCollection,
  activeDynamicBenefitGenerators = {
    // permet de déterminer les infos à générer pour les tests unitaires
    fsl: true,
    aidesVelo: true,
  },
  dynamicBenefitAttributes = additionalBenefitAttributes
)
```

On a une signature plus compréhensible, mais je garde les builders dans la fonction generate

Reply:
- Je trouve que la section qui enchaîne [les `if` est peu lisible](https://github.com/betagouv/aides-jeunes/compare/931-r%C3%A9%C3%A9crire-la-fonction-generate-des-aides...931-r%C3%A9%C3%A9crire-la-fonction-generate-des-aides-prop?expand=1#diff-ae5e253a9accf7ca88181b7ec3142017c3519b31611485aedd7c3b73e6c115f2L78-L87)
- Je trouve qu'il y a un bénéfice assez important à normaliser la forme d'[un générateur](https://github.com/betagouv/aides-jeunes/compare/931-r%C3%A9%C3%A9crire-la-fonction-generate-des-aides...931-r%C3%A9%C3%A9crire-la-fonction-generate-des-aides-prop?expand=1#diff-ae5e253a9accf7ca88181b7ec3142017c3519b31611485aedd7c3b73e6c115f2R82)